File: Leetcode/BST/220_Contains_Duplicate_III.cpp
class Solution {
public:
    bool containsNearbyAlmostDuplicate(vector<int>& nums, int k, int t) {
        if (nums.size()==0) return false;
        else if(k==0) return false;
        
        set<long long> myset;
        for(int i=0;i<nums.size();i++)
        {
        	if(i>k) myset.erase(nums[i-k-1]);

        	auto pos=myset.lower_bound((long long)nums[i]-t);
        	if(pos!=myset.end() && *pos-nums[i]<=t) return true;
        	myset.insert(nums[i]);
        }
        return false;
    }
};

File: Leetcode/BST/220_Contains_Duplicate_III.java
class Solution {
    public boolean containsNearbyAlmostDuplicate(int[] nums, int k, int t) {
        TreeSet<Long> set=new TreeSet<>();
        for(int i=0;i<nums.length;i++)
        {
        	Long floor=set.floor((long)nums[i]+t);
        	Long ceil=set.ceiling((long)nums[i]-t);

        	if( (floor!=null && floor >=nums[i]) || (ceil!=null && ceil<=nums[i]) )
        		return true;

    		set.add((long)nums[i]);
    		if(i>=k)
    		{
    			set.remove((long)nums[i-k]);
    		}
        }
        return false;
    }
}